src/storage/src/source/mod.rs

```diff
@@ -931,12 +943,384 @@ where
+                    message.headers,
+                )));
+            }
+            let ts_cap = cap.delayed(&timestamp);
+            output.session(&ts_cap).give_vec(&mut msg_output);
+        }
+
+        // We should now downgrade our capability.
+        let new_cap = self
+            .persisted_timestamp_bindings
+            .values()
+            .filter_map(|v| v.get(2))
+            .map(|(timestamp, _diff)| *timestamp)
+            .chain(std::iter::once(
+                *self.read_since.elements().first().unwrap_or(&u64::MAX),
+            ))
+            .min()
+            .unwrap_or_else(|| Timestamp::minimum());
+        let _ = cap.try_downgrade(&new_cap);
+
+        // Mint new timestamps
+        let new_ts = cap
+            .try_delayed(&(self.now)())
+            .map(|c| *c.time())
+            .unwrap_or(*cap.time());
+        let new_ac = Antichain::from_elem(new_ts + 1);
+        let append_updates: Vec<_> = self
+            .messages_by_part
+            .iter()
+            .filter_map(|(partition, messages)| {
+                let new_max_offset = messages
+                    .iter()
+                    .fold(0, |accum, msg| std::cmp::max(accum, msg.offset.offset));
+                let old_max_offset =
+                    if let Some(bindings) = self.persisted_timestamp_bindings.get(partition) {
+                        bindings.back().unwrap().1
+                    } else {
+                        0
+                    };
+                let offset_delta = new_max_offset - old_max_offset;
+                if offset_delta <= 0 {
+                    None
+                } else {
+                    Some(((&(), partition), new_ts, offset_delta))
+                }
+            })
+            .collect();
+        let cas_result = self
+            .write_handle
+            .compare_and_append(
+                Duration::from_secs(10),
+                append_updates.iter().map(|(kv, ts, diff)| (*kv, ts, diff)),
+                self.read_since.clone(),
+                new_ac.clone(),
+            )
+            .await
+            // TODO: handle these errors once they're refined
+            .expect("CAS first")
+            .expect("CAS second");
+
+        let pending_messages = self.messages_by_part.values().any(|m| !m.is_empty())
+            || self
+                .persisted_timestamp_bindings
+                .values()
+                .any(|m| m.len() > 1);
+
+        let source_status = match (cas_result, pending_messages, &self.processing_state) {
+            (_, _, SourceMessageProcessingState::Finished) => SourceStatus::Done,
+
+            (Err(_actual_upper), _, _) => {
+                self.activator.activate();
+                SourceStatus::Alive
+            }
+
+            // No CAS error, no pending messages, Draining -> draining is complete
+            (Ok(()), false, SourceMessageProcessingState::Draining) => {
+                let _ = self
```

Comment:
(sorry for the basic question but) how does it know when the source has no more incoming messages?

Reply:
`SourceReader::poll_next` returns `None` because we get `NextMessage::Finished`.  Importantly this doesn't happen for all source types! e.g. kafka is never Done.  Right now it's just s3 and file sources afaik

Reply:
aha! okay thank you